[#31] CI 구현을 위한 설정 추가

msugo1 opened the pull request:
### 📑  Jenkinsfile 추가

- `poll - build - test` 과정을 자동화 시키기 위한 `pipeline` 추가

### 📑  pom.xml

- Shell을 이용해 단위 테스트 자동실행을 하기 위한 `maven-surefire-plugin` 의존성 추가

- 통합 테스트는 `maven-failsafe-plugin`을 활용해 서버를 완전히 구축하기 전에도 CI가 실패하지 않도록 설정

### 📑 테스트 관련

- 전체 과정을 테스트 하는 통합 테스트의 경우 이름 앞에 `IT`를 붙여 `surefire`를 활용한 단위테스트에 포함되지 않도록 함

Commit:
[#31] CI 구현을 위한 설정 추가

* Jenkinsfile 추가

- `poll - build - test` 과정을 자동화 시키기 위한 `pipeline` 추가

* pom.xml

- Shell을 이용해 단위 테스트 자동실행을 하기 위한 `maven-surefire-plugin` 의존성 추가

- 통합 테스트는 `maven-failsafe-plugin`을 활용해 서버를 완전히 구축하기
  전에도 CI가 실패하지 않도록 설정

* 전체 과정을 테스트 하는 통합 테스트의 경우 이름 앞에 `IT`를 붙여
  `surefire`를 활용한 단위테스트에 포함되지 않도록 함

msugo1 commented:
![image](https://user-images.githubusercontent.com/68094005/111873548-1bf44f00-89d4-11eb-96b6-1589024ca308.png)

설정 수정해서 단위 테스트까지는 통과하게 만들었는데 이상하게 상태가 fail에서 안바뀌네요; Merge 같은 경우도 승인 전까지는 못바꾸게 제약 걸어두었습니다.

Commit:
[#31] 통합테스트 분리 및 CI PUSH 테스트

Commit:
[#31] 통합 테스트 재 분리 및 CI 테스트

f-lab-dev commented on pom.xml:
```diff
@@ -112,6 +112,18 @@
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.0.0-M5</version>
+            </plugin>
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <version>3.0.0-M5</version>
```
이 플러그인들은 어떤 목적의 플러그인일까요?

msugo1 replied:
### maven-surefire-plugin
- `surefire` 플러그인은 build의 life cycle 중에서 `test phase` 동안 사용되는 플러그인 입니다. 그리고 테스트를 마친 후 `.txt, .xml 포맷으로 테스트 결과를 제공`해주는 역할도 합니다. 
- 혹시 테스트 중 하나라도 실패하면 빌드가 즉시 빌드 실패의 결과를 실패로 변경합니다.

(https://maven.apache.org/surefire/maven-surefire-plugin/)

### maven-failsafe-plugin
- `failsafe` 플러그인도 test phase 을 위한 플러그인이지만 위의 sure-fire 플러그인과는 다르게 `통합테스트의 실행`을 위해 설계되었습니다.
-  이름이 failsafe으로 선정된 이유는 `1. surefire과 동의어` 그리고 `2.혹시 테스트가 실패하더라도 안전한 방식으로 실패` 시키기 때문이라고 합니다.  그래서 실제로 테스트 실행 시 테스트가 실패하더라도 빌드 실패와는 분리되어 결과가 나옵니다.
- `sure-fire`과 마찬가지로 테스트 리포트 결과를 제공해줍니다.

(https://maven.apache.org/surefire/maven-failsafe-plugin/)

Commit:
[#32] 친구에게 남긴 추천글을 수정할 기능 구현

Controller 계층 관련

* RecommendationController

- `leaveReview` 메소드가 요청을 반영하는데 성공하면
  `HttpStatus.CREATED`를 리턴하도록 변경

- 추천글 수정 요청을 받아서 서비스 계층에 전달해줄 `modifyReview` 메소드
  추가

Service 계층 관련

* RecommendationService

- 들어온 수정 요청을 가지고 검증과정을 거쳐 `검증을 통과한 요청`의 경우는
  Mapper로 전달해 줄 `modifyRecommendation` 메소드 추가

- 추천글이 `남겨진 시점으로부터의 날짜`를 반환해 줄 `howLongSinceWrittenAt` 메소드 추가

- 추천글이 `수정가능한 기간 내에 있는지 검증`해 줄
  `validateModificationAvailability` 메소드 추가

- 추천글을 남길 수 있는지 검증해주는 메소드의 이름을
  `validateLeaveRecommendationAvailability`로 변경

Mapper 계층 관련

* RecommendationMapper

- DB에서 `현재 시점에서 추천글이 남겨진 시점의 날짜 차이`가 얼마나 되는지
  계산해서 리턴해 줄 `getHowLongSinceWrittenAt` 메소드 추가 (Optional을
  사용해서 NPE로 부터 안전하도록 변경)

- 서비스 계층에서 검증과정을 모두 통과한 추천글 수정요청을 DB에 반영할
  수 있도록하는 `updateRecommendation` 메소드 추가

* RecommendationMapper.xml

- 위의 요청을 처리해 줄 쿼리문 추가

Model 관련

* RecommendationRequest

- `빈 객체만 만들어지는 불상사`를 막기 위해 NoArgsConstructor에 접근
  제어자를 `private`로 설정

테스트 관련

- 추천글 수정 요청 처리에 대한 비즈니스 로직을 검증할 단위 테스트 작성

- 포스트맨을 활용해 전반적인 테스트 및 추가한 메소드 동작 확인

Commit:
[#33] 페이징 방식 변경

* `offset` 방식의 페이징에서 대용량 데이터를 처리할 때 더 큰 장점을 가진  `cursor` 방식의 페이징으로 변경

Commit:
[#31] Jenkinsfile 파이프 라인 변경

- 현재 서버 연결 혹은 application.properties를 외부에서 주입 받도록
  설정이 되어있지 않으므로 차후에 추가하고자 통합테스트 부분 삭제

msugo1 committed:
Merge pull request #73 from f-lab-edu/feature/32

[#32] 친구에게 남긴 추천글을 수정할 기능 구현

msugo1 committed:
Merge pull request #74 from f-lab-edu/feature/33

[#33] 페이징 방식 변경

msugo1 committed:
Merge branch 'develop' into feature/31
